### ISSUE TITLE
packet-solutions/xdp_prog_kern_03.c/SEC(xdp_icmp_echo) offset error

### ISSUE THREAD

**Issue by yugun** [xdp-project/xdp-tutorial#198]
Hi

I encountered offset error while inserting xdp_icmp_echo.

mount -t bpf bpf /sys/fs/bpf
./xdp_loader -d eth0 -F --filename xdp_prog_kern_03.o --progsec xdp_icmp_echo

the error message is as follows

> ... omitted ...
> 
> 81: (55) if r5 != 0x0 goto pc+95
>  R0_w=inv0 R1=pkt(id=0,off=0,r=42,imm=0) R2=pkt(id=0,off=22,r=42,imm=0) R3=inv2 R4=inv128 R5_w=inv0 R6=ctx(id=0,off=0,imm=0) R7=pkt(id=1,off=22,r=26,umin_value=8,umax_value=60,var_off=(0x0; 0x3c)) R8=inv1 R10=fp0
> 82: (61) r3 = *(u32 *)(r2 +20)
> **invalid access to packet, off=42 size=4, R2(id=0,off=22,r=42)**
> R2 offset is outside of the packet
> processed 177 insns (limit 1000000) max_states_per_insn 1 total_states 10 peak_states 10 mark_read 5
> 
> libbpf: -- END LOG --
> libbpf: failed to load program 'xdp_icmp_echo'
> libbpf: failed to load object 'xdp_prog_kern_03.o'
> ERR: loading BPF-OBJ file(xdp_prog_kern_03.o) (-22): Invalid argument
> ERR: loading file: xdp_prog_kern_03.o

I  keep commenting out some code lines in order to find where the problem is.
finally, i got no error when _swap_src_dst_ipv6_ is hidden.
And I modified code like follows.

```
...
else if (eth_type == bpf_htons(ETH_P_IPV6)
           && ip_type == IPPROTO_ICMPV6
           && icmp_type == ICMPV6_ECHO_REQUEST) {
      swap_src_dst_ipv6(ipv6hdr);
      echo_reply = ICMPV6_ECHO_REPLY;
}
...
```

And my question as follows
1. Doesn't  _parse_ipv6hdr_ guarantee _swap_src_dst_ipv6_ ?
or Is my kernel's verifier weird?

2. Is there easy way to find where the problem is? 

Thank you !

**Comment by tohojo** [comment]
Yugun Kang <notifications@github.com> writes:

> Hi
>
> I encountered offset error while inserting xdp_icmp_echo.
>
> mount -t bpf bpf /sys/fs/bpf
> ./xdp_loader -d eth0 -F --filename xdp_prog_kern_03.o --progsec xdp_icmp_echo
>
> the error message is as follows
>
>> ... omitted ...
>> 
>> 81: (55) if r5 != 0x0 goto pc+95
>>  R0_w=inv0 R1=pkt(id=0,off=0,r=42,imm=0) R2=pkt(id=0,off=22,r=42,imm=0) R3=inv2 R4=inv128 R5_w=inv0 R6=ctx(id=0,off=0,imm=0) R7=pkt(id=1,off=22,r=26,umin_value=8,umax_value=60,var_off=(0x0; 0x3c)) R8=inv1 R10=fp0
>> 82: (61) r3 = *(u32 *)(r2 +20)
>> **invalid access to packet, off=42 size=4, R2(id=0,off=22,r=42)**
>> R2 offset is outside of the packet
>> processed 177 insns (limit 1000000) max_states_per_insn 1 total_states 10 peak_states 10 mark_read 5
>> 
>> libbpf: -- END LOG --
>> libbpf: failed to load program 'xdp_icmp_echo'
>> libbpf: failed to load object 'xdp_prog_kern_03.o'
>> ERR: loading BPF-OBJ file(xdp_prog_kern_03.o) (-22): Invalid argument
>> ERR: loading file: xdp_prog_kern_03.o
>
> I  keep commenting out some code lines in order to find where the problem is.
> finally, i got no error when _swap_src_dst_ipv6_ is hidden.
> And I modified code like follows.
>
> ```
> ...
> else if (eth_type == bpf_htons(ETH_P_IPV6)
>            && ip_type == IPPROTO_ICMPV6
>            && icmp_type == ICMPV6_ECHO_REQUEST) {
>       swap_src_dst_ipv6(ipv6hdr);
>       echo_reply = ICMPV6_ECHO_REPLY;
> }
> ...
> ```
>
> And my question as follows
> 1. Doesn't  _parse_ipv6hdr_ guarantee _swap_src_dst_ipv6_ ?
> or Is my kernel's verifier weird?

Well, this does seem like an error that shouldn't occur (and indeed it
doesn't for me). Which kernel and LLVM versions are you using?

> 2. Is there easy way to find where the problem is?

Well, the verifier is telling you:

> **invalid access to packet, off=42 size=4, R2(id=0,off=22,r=42)**

so it's saying it can't prove that the program is checking that a 4-byte
access to the packet at offset 42 is not outside the packet bounds.

**Comment by yugun** [comment]
Thank you for your reply

> Well, this does seem like an error that shouldn't occur (and indeed it
> doesn't for me). Which kernel and LLVM versions are you using?

My kernel version is 5.4.82 and LLVM is 11.0.0.

**Comment by tohojo** [comment]
Yugun Kang <notifications@github.com> writes:

> Thank you for your reply
>
>> Well, this does seem like an error that shouldn't occur (and indeed it
>> doesn't for me). Which kernel and LLVM versions are you using?
>
> My kernel version is 5.4.82 and LLVM is 11.0.0.

Hmm, my bet would be that the verifier has simply gotten smarter, then.
Any chance you could test on a newer kernel?

**Comment by kkh913** [comment]
Hello?

I came here while googled for the same problem as yours.

I have solved this problem and would like to inform you.

# TL;DR

| use clang-10 not clang
 
go to common/common.mk, and modify ... 

``` CLANG ?= clang ```

to

``` CLANG ?= clang-10 ```


# Let's deep dive

This is a sad situation. Why does clang-11 point to the ipv6 saddr pointer to +20 instead of +8?

| 82: (61) r3 = *(u32 *)(r2 +20)

I modified the if statements and pointers in various ways, but it didn't help at all.

Now, I'm running this tutorial on Ubuntu 20.10. The kernel version is 5.8.0-45-generic.

And the installation status of clang related packages is as follows:
```
➜ apt list --installed | grep clang

WARNING: apt does not have a stable CLI interface. Use with caution in scripts.

clang-10/groovy,now 1:10.0.1-6 amd64 [installed]
clang-11/groovy,now 1:11.0.0-2 amd64 [installed,automatic]
clang-9/groovy,now 1:9.0.1-14build1 amd64 [installed]
clang-format-11/groovy,now 1:11.0.0-2 amd64 [installed,automatic]
clang-format/groovy,now 1:11.0-51~exp1 amd64 [installed]
clang/groovy,now 1:11.0-51~exp1 amd64 [installed]
libclang-11-dev/groovy,now 1:11.0.0-2 amd64 [installed,automatic]
libclang-9-dev/groovy,now 1:9.0.1-14build1 amd64 [installed]
libclang-common-10-dev/groovy,now 1:10.0.1-6 amd64 [installed,automatic]
libclang-common-11-dev/groovy,now 1:11.0.0-2 amd64 [installed,automatic]
libclang-common-9-dev/groovy,now 1:9.0.1-14build1 amd64 [installed,automatic]
libclang-cpp10/groovy,now 1:10.0.1-6 amd64 [installed,automatic]
libclang-cpp11/groovy,now 1:11.0.0-2 amd64 [installed,automatic]
libclang-cpp9/groovy,now 1:9.0.1-14build1 amd64 [installed,automatic]
libclang-dev/groovy,now 1:11.0-51~exp1 amd64 [installed]
libclang1-10/groovy,now 1:10.0.1-6 amd64 [installed,automatic]
libclang1-11/groovy,now 1:11.0.0-2 amd64 [installed,automatic]
libclang1-9/groovy,now 1:9.0.1-14build1 amd64 [installed,automatic]
```

clang-10 was not installed intentionally, but it was installed. 

As mentioned in TL;DR, when the clang version was replaced, it works without any problems.

Please "closed" if it helped.

**Comment by yugun** [comment]
Thank you for suggesting a solution.

But, It's hard to do it right away.

I have installed clang-11.0 using source code and I don't have clang-10.0.

I'll leave it to the manager to close this issue.

Thank you again
